### PR TITLE
sctest: refactor cumulative tests

### LINKED ISSUE #106920: schemachanger,sctest: backup tests are slightly broken
In an effort to understand how the `TestBackup*` cumulative tests for the declarative schema changer work, I noticed a subtle bug: we back up _after_ executing each post-commit stage, which is fine but means we never test backups after the stmt txn commits but before the job runs, which is arguably the most important case.

~As it turns out, fixing this reveals other longstanding bugs.~

This issue tracks the need to refactor the `sctest` framework, especially the backups tests.

Jira issue: CRDB-29767

### PR DESCRIPTION
**Pull request by postamar**
This commit comprehensively refactors the cumulative test definitions for the declarative schema changer.

Fixes #106920.

Release note: None

### CODE REVIEW

File: pkg/ccl/schemachangerccl/BUILD.bazel [hunk 1]
```diff
@@ -93,34 +79,22 @@ sctest_gen(
...
-    new_cluster_func = "newMultiRegionCluster",
+    new_cluster_factory = "MultiRegionTestClusterFactory{}",
     package = "schemachangerccl",
     suffix = "_ccl",
     test_data = [":end_to_end_ccl_testdata_definitions"],
     tests = [
         "Backup",
+        "BackupMixedVersion",
         "EndToEndSideEffects",
+        "ExecuteWithDMLInjection",
```
Comment (postamar) [inline pkg/ccl/schemachangerccl/BUILD.bazel:90]: `ExecuteWithDMLInjection ` tests were missing in CCL for no good reason.

File: pkg/ccl/schemachangerccl/BUILD.bazel [hunk 2]
```diff
@@ -93,34 +79,22 @@ sctest_gen(
...
         "GenerateSchemaChangeCorpus",
         "Pause",
         "Rollback",
+        "ValidateMixedVersionElements",
```
Comment (postamar) [inline pkg/ccl/schemachangerccl/BUILD.bazel:94]: `ValidateMixedVersionElements` tests were missing in CCL for no good reason.

File: pkg/ccl/schemachangerccl/testdata/end_to_end/create_index/create_index.definition [hunk 3]
```diff
@@ -1,43 +1,43 @@
...
-INSERT INTO t VALUES($stageKey + 2, 'c');
-DELETE FROM t WHERE v = 'a' and k=$stageKey;
-INSERT INTO t VALUES($stageKey, 'a');
-UPDATE t SET v='a' WHERE k > 0;
+INSERT INTO t VALUES ($stageKey, 'a');
+INSERT INTO t VALUES ($stageKey + 1, 'b');
+INSERT INTO t VALUES ($stageKey + 2, 'c');
+DELETE FROM t WHERE name = 'a' and id = $stageKey;
+INSERT INTO t VALUES ($stageKey, 'a');
+UPDATE t SET name = 'a' WHERE id > 0;
```
Comment (postamar) [inline pkg/ccl/schemachangerccl/testdata/end_to_end/create_index/create_index.definition:11]: This whole test definition needed some love in order to pass the `ExecuteWithDMLInjection` test.

File: pkg/sql/schemachanger/sctest/test_server_factory.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	// will be the minimum supported binary version.
+	WithMixedVersion() TestServerFactory
+
+	// Run creates a test cluster and applies fn to it.
+	Run(
+		ctx context.Context,
+		t *testing.T,
+		fn func(s serverutils.TestServerInterface, tdb *gosql.DB),
+	)
+}
```
Comment (postamar) [inline pkg/sql/schemachanger/sctest/test_server_factory.go:50]: `TestServerFactory` and its vanilla and multi-region implementations replace the old `newClusterFunc`.

**Review by postamar — COMMENTED** [review]
Review guide: at some point, this PR is going to require a leap of faith from the reviewer. Personally, I'm cool with that because I'm not convinced that the existing tests were that well understood in the first place.

All of the pre-existing tests have been either maintained or fixed. Test cases are no longer randomly skipped.

**Review by fqazi — APPROVED** [review]
Excellent work this is pretty clean! :LGTM:

> Reviewed 29 of 33 files at r1, 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/106743#-:-N_uEGTvEXX_nuv4L2i3:b-oyhrya)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @postamar)

___
*[`pkg/sql/schemachanger/sctest/backup.go` line 411 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/106743#-N_uCcCvAc9JFTAVXd2i:-N_uCcCvAc9JFTAVXd2j:bguou2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/fedb6d58f6c699be24022194280b56a7c2f40c90/pkg/sql/schemachanger/sctest/backup.go#L411)):*
> ```Go
> 			// Clean up.
> 			// Hack to deal with corrupt userfiles tables due to #76764.
> 			const vq = `SELECT * FROM crdb_internal.invalid_objects WHERE database_name != 'backups'`
> ```

Can we clean this up now?
___
*[`pkg/ccl/schemachangerccl/testdata/end_to_end/create_index/create_index.definition` line 11 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/106743#gh-1270207398:-N_uDKCz46c9calEqYmL:b6zrrk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b712157fd8208b8b1848b2d4dd2b5b61a050fb3c/pkg/ccl/schemachangerccl/testdata/end_to_end/create_index/create_index.definition#L11)):*
<details><summary><i>Previously, postamar (Marius Posta) wrote…</i></summary><blockquote>

This whole test definition needed some love in order to pass the `ExecuteWithDMLInjection` test.
</blockquote></details>

Good catch!

**Comment by postamar** [comment]
Thanks for the review! There's still one outstanding issue to resolve in this PR in that the `schemachangerccl` now has an unholy amount of tests. I got rid of the randomized skipping because I didn't like it. I'd much rather we deliberately skip entire classes of checks in regular CI and skip nothing during the nightlies. For instance, in the backup tests, skip everything except RESTORE DATABASE cases for successful schema changes in non-mixed clusters. I'm not sure how to do this properly.

> Can we clean this up now?

I hadn't even looked into this. Presumably! I'll give it a whirl.

**Comment by postamar** [comment]
I'm battling the `maybe_stress` portion of the CI. See https://cockroachlabs.slack.com/archives/CJ0H8Q97C/p1690203313338359 for details.

**Comment by fqazi** [comment]
> @fqazi also I renamed the `TestValidateMixedVersionElements` to `TestPauseMixedVersion` because from what I can tell it's a mixed-version flavour of `TestPause`.
That sounds good, and its a less verbose description.

> I got rid of the randomized skipping because I didn't like it. I'd much rather we deliberately skip entire classes of checks in regular CI and skip nothing during the nightlies.
I think that's a good approach and I'm guessing local builds also run everything too

**Pull request merged**